planetiler-core/src/main/java/com/onthegomap/planetiler/expression/MultiExpression.java

```diff
@@ -220,38 +225,38 @@ public boolean isEmpty() {
     private final Map<String, List<EntryWithId<T>>> keyToExpressionsMap;
     // same as keyToExpressionsMap but as a list (optimized for iteration when # source feature keys > # tags we care about)
     private final List<Map.Entry<String, List<EntryWithId<T>>>> keyToExpressionsList;
-    // expressions that should match when certain tags are *not* present on an input element
-    private final List<Map.Entry<String, List<EntryWithId<T>>>> missingKeyToExpressionList;
-    // expressions that match a constant true input element
-    private final List<EntryWithId<T>> constantTrueExpressionList;
+    // expressions that must always be evaluated on each input element
+    private final List<EntryWithId<T>> alwaysEvaluateExpressionList;
 
     private KeyIndex(MultiExpression<T> expressions) {
       int id = 1;
       // build the indexes
       Map<String, Set<EntryWithId<T>>> keyToExpressions = new HashMap<>();
-      Map<String, Set<EntryWithId<T>>> missingKeyToExpressions = new HashMap<>();
-      List<EntryWithId<T>> constants = new ArrayList<>();
+      List<EntryWithId<T>> always = new ArrayList<>();
 
       for (var entry : expressions.expressions) {
         Expression expression = entry.expression;
         EntryWithId<T> expressionValue = new EntryWithId<>(entry.result, expression, id++);
-        getRelevantKeys(expression,
-          key -> keyToExpressions.computeIfAbsent(key, k -> new HashSet<>()).add(expressionValue));
-        getRelevantMissingKeys(expression,
-          key -> missingKeyToExpressions.computeIfAbsent(key, k -> new HashSet<>()).add(expressionValue));
-        if (expression.equals(TRUE)) {
-          constants.add(expressionValue);
+        if (mustAlwaysEvaluate(expression)) {
+          always.add(expressionValue);
+        } else {
+          getRelevantKeys(expression,
+            key -> keyToExpressions.computeIfAbsent(key, k -> new HashSet<>()).add(expressionValue));
         }
       }
       // create immutable copies for fast iteration at matching time
-      constantTrueExpressionList = List.copyOf(constants);
+      if (!always.isEmpty()) {
+        LOGGER.warn("{} expressions will be evaluated for every element:", always.size());
+        for (var expression : always) {
+          LOGGER.warn("    {}", expression.expression.generateJavaCode());
```

Comment:
Fix should be to surround the warn call with `if(LOGGER.isWarnEnabled()) {`

Reply:
There are a few cases like this that I think I want to tackle in one pass (and leave highlighted so I don't miss one). I'd prefer to go the route of a logger framework that takes a lambda that only gets invoked if that level is needed. There's also some other logging issues to clear up like `core` should just depend on slf4j and the runnable thing (`dist` module or a project that depends on this) should bind slf4j to whatever implementation it wants.

Reply:
Good plan.  Looks like https://github.com/kwon37xi/slf4j-lambda exists

Reply:
also slf4j v2 api will support lambdas. I'd like to use that but it's been in alpha stage since 2019...? https://mvnrepository.com/artifact/org.slf4j/slf4j-api

Reply:
It's a logging framework, how bad could alpha be?  🤔